### ISSUE TITLE
Whitelist not applied

### ISSUE THREAD

**Issue by mirecalu** [arturbosch/detekt#1589]
I'm trying to add some rules into the Whitelist but it has no effect on the report file.
There is an example of `UndocumentedPublicClass` in `MenuEntryView.kt` that should be white listed.

```kotlin
//detekt.gradle
detekt {
    config = files("$rootDir/default-detekt-config.yml")
    filters = ".*build.*,.*/resources/.*,.*/tmp/.*"
    baseline = file("$rootDir/detekt-baseline.xml")
}
```

```xml
<!--detekt-baseline.xml-->
<?xml version="1.0" ?>
<SmellBaseline>
    <Whitelist>
        <ID>
            UndocumentedPublicClass:MenuEntryView.kt$com.foo.bar.menu.MenuEntryView.kt
        </ID>
    </Whitelist>
</SmellBaseline>
```
I don't manage to see what I'm missing into the documentation

**Comment by arturbosch** [comment]
Hm can you try to inline the the ID like `<ID>[issue]</ID>`?
Maybe its the whitespace.

**Comment by mirecalu** [comment]
Yes, It didnt work either

**Comment by marschwar** [comment]
Can you please run `./gradlew detektBaseline` to create a baseline file. I did just that and received

```
<?xml version="1.0" ?>
<SmellBaseline>
  <Blacklist timestamp="1555415294513"></Blacklist>
  <Whitelist timestamp="1555415294513">
    <ID>UndocumentedPublicClass:MenuEntryView.kt$MenuEntryView</ID>
  </Whitelist>
</SmellBaseline>
```

As you can see, the simple class name instead of the fully qualified name is used.

**Comment by mirecalu** [comment]
When I run `./gradlew detektBaseline` I only have theses two empty list (even if I have plenty of errors in the HTML report when I run `./gradlew detekt`)
```xml
<?xml version="1.0" ?>
<SmellBaseline>
  <Blacklist timestamp="1555432176090"></Blacklist>
  <Whitelist timestamp="1555432176857"></Whitelist>
</SmellBaseline>
```

I also tried to add this line ` <ID>UndocumentedPublicClass:MenuEntryView.kt$MenuEntryView</ID>`, it doesnt work either

**Comment by marschwar** [comment]
Which version of detekt are you using?

**Comment by mirecalu** [comment]
Okey I figured it out ! 
I have two modules in my AndroidProject : 
- app : which has several findings in detekt
- cleanModule ; which has NO finding

When I ran ./gradlew detektBaseline, there was two tasks : 
`app:detektBaseline` -> detekt-baseline.xml was created with some items in the whitelist
`cleanModule:detektBaseline` -> detekt-baseline.xml is override whitout `app` whitelist

Is it possible to append detekt issue when there is several modules in our app ?

**Comment by arturbosch** [comment]
Unfortunately not.

One possibility to implement this for the gradle plugin would be to check if the baseline file already exist and merge it's findings with the new findings (with filtering for duplicates).

**Comment by arturbosch** [comment]
From a discussion on slack:
```
what am I missing? how can I have a centralized baseline for the project?
Artur Bosch  30 days ago
Unfortunately we haven't figured out how to have a centralized baseline file across different gradle modules.
Something that works for Gradle but is also as generic to be included into detekt-api would be great.
Artur Bosch  30 days ago
An idea which just came up:
we could have a modularized baseline which have different sections for each gradle module like detekt-cli and just override detekt-cli when run ...
Artur Bosch  30 days ago
The generic part is tricky... we could add a --baseline [prefix:]path option which the gradle plugin could use to let detekt know which baseline part to override...
```

**Comment by kenyee** [comment]
Would indeed be nice if it did this:

> One possibility to implement this for the gradle plugin would be to check if the baseline file already exist and merge it's findings with the new findings (with filtering for duplicates).

Otherwise, anyone have nice scripts that do this since Detekt can only work on a single file?

**Comment by ZacSweers** [comment]
I wrote a very simple one internally to batch and merge. One problematic bit is that the detekt plugin currently reuses the `baseline` extension parameter for both inputs to some tasks and outputs of baseline creation tasks. I think it would be wise to separate these to properly support this. Happy to upstream this via PR

```kotlin
/**
 * Collects all generated baselines and merges them into one global one. Temporary until detekt supports this as a first
 * party tool: https://github.com/arturbosch/detekt/issues/1589.
 *
 * This is only configured to run when the [TASK_NAME] task name is passed in, as detekt currently reuses the
 * `baseline` argument for both the input file to regular detekt tasks and output file of its create tasks. Since we
 * don't want the create tasks to overwrite each other into the same output task, we dynamically configure this as
 * needed. When [TASK_NAME] is specified, all detekt baselines are pointed to an intermediate output file in that
 * project's build directory and the misc "detektBaseline" tasks are wired to have their outputs to be inputs to this
 * task's [baselineFiles].
 */
@CacheableTask
abstract class MergeDetektBaselinesTask : DefaultTask() {

  companion object {
    const val TASK_NAME = "detektBaselineMerge"
  }

  init {
    description = " Collects all generated detekt baselines and merges them into one global one."
    group = LifecycleBasePlugin.VERIFICATION_GROUP
  }

  @get:PathSensitive(PathSensitivity.RELATIVE)
  @get:InputFiles
  abstract val baselineFiles: ConfigurableFileCollection

  @get:OutputFile
  abstract val outputFile: RegularFileProperty

  @TaskAction
  fun merge() {
    val format = BaselineFormat()
    val merged = baselineFiles
        .filter { it.exists() }
        .map { format.read(it.toPath()) }
        .reduce { acc, baseline ->
          acc.copy(
              blacklist = acc.blacklist.copy(acc.blacklist.ids + baseline.blacklist.ids),
              whitelist = acc.whitelist.copy(acc.whitelist.ids + baseline.whitelist.ids)
          )
        }
    val sortedAndDedupedMerge = merged.copy(
        blacklist = merged.blacklist.copy(merged.blacklist.ids.toSortedSet()),
        whitelist = merged.whitelist.copy(merged.whitelist.ids.toSortedSet())
    )

    format.write(sortedAndDedupedMerge, outputFile.asFile.get().toPath())
  }
}
```

combined with

```kotlin
// root project
val mergeDetektBaselinesTask = ...

// subprojects...
// Internally we gate configuring this based on whether or not the "mergeDetektBaselines" task was invoked, due to the mentioned awkward configuration of the baseline argument
tasks.withType<DetektCreateBaselineTask>().configureEach {
  mergeDetektBaselinesTask.configure {
    baselineFiles.from(baseline)
  }
}
file("$buildDir/intermediates/detekt/baseline.xml")
```

**Comment by BraisGabin** [comment]
I found this: https://developer.android.com/studio/test/command-line#multi-module-reports

Android created it's own plugin to merge the output of their reports. Maybe that's what detekt should do too.

**Comment by ZacSweers** [comment]
That's effectively what we did internally in our own plugin. If there's interest I could look at contributing it

**Comment by r4zzz4k** [comment]
That would actually be super awesome! This issue was a major blocker for a team on one of my previous projects, never had a chance to look into it :(

**Comment by BraisGabin** [comment]
> That's effectively what we did internally in our own plugin. If there's interest I could look at contributing it

To be honest I think that detekt have far too much baselines (the "plain old one", main, test, debug, release, all the flavours that you want to have, js, iOS...) and I don't know if we should merge all of them together or just to merge them by "type" or if there is even a correct answer here. And when I think about unified reports (related problem) it seems an even more difficult question. Probably the problem is not to have multiple baselines/reports but to have multiple runs...

Any way, @ZacSweers if you can contribute that it would help to see what the users need exactly :).

**Comment by mrclrchtr** [comment]
@ZacSweers unfortunately the code snippet above can't be used anymore, because BaselineFormat() doesn't seem to be public anymore. If you have a solution, I would be very happy if you could share it.
